cdap-app-fabric/src/main/java/io/cdap/cdap/internal/namespace/credential/DefaultNamespaceCredentialProviderService.java

```diff
@@ -0,0 +1,149 @@
+ */
+public class DefaultNamespaceCredentialProviderService extends AbstractIdleService
+    implements NamespaceCredentialProviderService {
+
+  private static final String CREDENTIAL_PROVIDER_NAME = "gcp-wi-credential-provider";
+  private final CredentialProvider credentialProvider;
+  private final NamespaceAdmin namespaceAdmin;
+  private final ContextAccessEnforcer contextAccessEnforcer;
+  private final CredentialProfileManager credentialProfileManager;
+  private final CConfiguration cConf;
+
+
+  @Inject
+  DefaultNamespaceCredentialProviderService(CConfiguration cConf,
+      CredentialProvider credentialProvider, ContextAccessEnforcer contextAccessEnforcer,
+      NamespaceAdmin namespaceAdmin, CredentialProfileManager credentialProfileManager) {
+    this.cConf = cConf;
+    this.credentialProvider = credentialProvider;
+    this.contextAccessEnforcer = contextAccessEnforcer;
+    this.namespaceAdmin = namespaceAdmin;
+    this.credentialProfileManager = credentialProfileManager;
+  }
+
+  /**
+   * Provisions a short-lived credential for the provided identity using the provided identity.
+   *
+   * @param namespace The identity namespace.
+   * @param scopes A comma separated list of OAuth scopes requested.
+   * @return A short-lived credential.
+   * @throws CredentialProvisioningException If provisioning the credential fails.
+   * @throws IOException If any transport errors occur.
+   * @throws NotFoundException If the profile or identity are not found.
+   */
+  @Override
+  public ProvisionedCredential provision(String namespace, String scopes)
+      throws CredentialProvisioningException, IOException, NotFoundException {
+    // contextAccessEnforcer.enforce(new NamespaceId(namespace),
+    //     NamespacePermission.PROVISION_CREDENTIAL);
+    NamespaceMeta namespaceMeta;
+    try {
+      namespaceMeta = namespaceAdmin.get(new NamespaceId(namespace));
+    } catch (Exception e) {
+      throw new IOException(String.format("Failed to get namespace '%s' metadata",
+          namespace), e);
+    }
+    String identityName =
+        GcpWorkloadIdentityUtil.getWorkloadIdentityName(namespaceMeta.getIdentity());
+    switchToInternalUser();
+    return credentialProvider.provision(namespace, identityName, scopes);
+  }
+
+  private void switchToInternalUser() {
+    SecurityRequestContext.reset();
+  }
+
+  /**
+   * Start the service.
+   */
+  @Override
+  protected void startUp() throws Exception {
+    // Updates the namespaces and creates the identities if not exist.
+    List<NamespaceMeta> namespaceMetaList = namespaceAdmin.list();
+    for (NamespaceMeta namespaceMeta : namespaceMetaList) {
+      if (Strings.isNullOrEmpty(namespaceMeta.getIdentity())) {
+        String identityName = namespaceAdmin.getIdentity(namespaceMeta.getNamespaceId());
+        MasterEnvironment masterEnv = MasterEnvironments.getMasterEnvironment();
+        if (masterEnv != null) {
```

Comment:
nit: we can make this a bit more readable by avoiding double nested conditionals and using &&.

Reply:
done